python/uv/repositories.bzl

```diff
@@ -0,0 +1,122 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+EXPERIMENTAL: This is experimental and may be removed without notice
+
+Create repositories for uv toolchain dependencies
+"""
+
+load("@bazel_tools//tools/build_defs/repo:utils.bzl", "maybe")
+load("//python/uv/private:toolchain_types.bzl", "UV_TOOLCHAIN_TYPE")
+load("//python/uv/private:toolchains_repo.bzl", "uv_toolchains_repo")
+load("//python/uv/private:versions.bzl", "UV_PLATFORMS", "UV_TOOL_VERSIONS")
+
+UV_BUILD_TMPL = """\
+# Generated by repositories.bzl
+load("@rules_python//python/uv:toolchain.bzl", "uv_toolchain")
+
+uv_toolchain(
+    name = "uv_toolchain",
+    uv = "{binary}",
+    version = "{version}",
+)
+"""
+
+def _uv_repo_impl(repository_ctx):
+    platform = repository_ctx.attr.platform
+    uv_version = repository_ctx.attr.uv_version
+
+    is_windows = "windows" in platform
+
+    suffix = ".zip" if is_windows else ".tar.gz"
+    filename = "uv-{platform}{suffix}".format(
+        platform = platform,
+        suffix = suffix,
+    )
+    url = "https://github.com/astral-sh/uv/releases/download/{version}/{filename}".format(
+        version = uv_version,
+        filename = filename,
+    )
+    if filename.endswith(".tar.gz"):
+        strip_prefix = filename[:-len(".tar.gz")]
+    else:
+        strip_prefix = ""
+
+    repository_ctx.download_and_extract(
+        url = url,
+        sha256 = UV_TOOL_VERSIONS[repository_ctx.attr.uv_version][repository_ctx.attr.platform].sha256,
+        stripPrefix = strip_prefix,
+    )
+
+    binary = "uv.exe" if is_windows else "uv"
+    repository_ctx.file(
+        "BUILD.bazel",
+        UV_BUILD_TMPL.format(
+            binary = binary,
+            version = uv_version,
+        ),
+    )
+
+uv_repository = repository_rule(
+    _uv_repo_impl,
+    doc = "Fetch external tools needed for uv toolchain",
+    attrs = {
+        "platform": attr.string(mandatory = True, values = UV_PLATFORMS.keys()),
+        "uv_version": attr.string(mandatory = True, values = UV_TOOL_VERSIONS.keys()),
+    },
+)
+
+# buildifier: disable=unnamed-macro
+def uv_register_toolchains(uv_version = None, register_toolchains = True):
+    """Convenience macro which does typical toolchain setup
+
+    Skip this macro if you need more control over the toolchain setup.
+
+    Args:
+        uv_version: The uv toolchain version to download.
+        register_toolchains: If true, repositories will be generated to produce and register `uv_toolchain` targets.
+    """
+    if not uv_version:
+        fail("uv_version is required")
+
+    toolchain_names = []
+    toolchain_labels_by_toolchain = {}
+    toolchain_compatible_with_by_toolchain = {}
+
+    for platform in UV_PLATFORMS.keys():
+        uv_repository_name = UV_PLATFORMS[platform].default_repo_name
+
+        maybe(
+            uv_repository,
```

Comment:
Note, that in `bzlmod` `maybe` does not really work. If you try to call `uv_register_toolchains` twice in `bzlmod`, this will just fail, so it might be good to just not use it.

Reply:
Oh TIL. I must be really lucky that I've not run into this yet. I'll remove it.

If there a link or reference to learn about this? Or is it just that with bzlmod "maybe" doesn't make sense because there is a DAG to plan downloads somehow?

Reply:
More the latter. In bzlmod, Bazel has the DAG of modules, and then the modules create repositories. Unlike workspace, the repo names aren't global. Instead, they're "scoped" to a particular bzlmod extension. When the module "rules_python" bzlmod extension "uv" creates a repo named "stuff", it gets a secret global (called "canonical") name like "rules_python~0.34.0~uv~stuff".

This is why `use_repo(uv, "stuff")` comes into play: it brings the "stuff" name into the module's "scope" so that Bazel understands that "@stuff" refers to rules_python's uv's repo.

So yeah, you can just omit using maybe entirely in bzlmod.

Reply:
Addressed in: https://github.com/bazelbuild/rules_python/pull/2062